[2단계 - 출석 다시 구현하기] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3898


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- 1단계에서 지키지 못했던 부분을 개선하려고 노력하여 점수를 상향했습니다.
  - 모든 원시값과 문자열을 포장하려고 노력했습니다.
  - 모든 엔티티를 작게 유지하려고 노력했습니다.
  - getter를 최대한 사용하지 않으려고 노력했습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 샐리! 2단계 미션도 잘 부탁드립니다.😊
1단계 미션에서 마지막에 코멘트 주셨던 부분은 DM으로 이야기 나눴던 것처럼 [1단계 브랜치](https://github.com/ljhee92/java-attendance/commits/step1/)에서 확인해주시면 감사드리겠습니다!
- [1단계 PR 링크](https://github.com/woowacourse/java-attendance/pull/36)
- [PR 이후 수정 시작한 커밋](https://github.com/ljhee92/java-attendance/commit/cea464b5f71b1ab7c00ad51464d9ff56f4c5c27e)

### commit 단위
- 1단계에서 페어와 진행했던 TDD보다 훨씬 엄격하게 적용해보았습니다.
- red -> green -> refactor 순으로 커밋을 나누었는데, 의식적으로 커밋하는 연습도 되고 TDD 싸이클을 익힐 수 있어서 개인적으로는 도움이 되었습니다.
- 다만 커밋 내역이 너무 많아지는 건 아닌지, TDD를 연습하는 동안에는 괜찮은건지 궁금합니다.
- 물론 회바회 팀바팀이겠지만 실무에서는 TDD를 적용하더라도 기능 단위로 커밋을 하는지 red -> green -> refactor 싸이클을 따라 커밋하는지도 궁금합니다.

### TDD, Top -> Down? Bottom -> Up?
- 1단계와 마찬가지로 저는 가장 중요한 도메인인 ``Attendance`` 먼저 테스트 코드를 작성하고, 프로덕션 코드를 구현했습니다.(Top -> Down 방식으로 TDD 진행)
- 그러다 보니 더 작은 단위로 볼 수 있는 캠퍼스 운영 시간, 교육 시간을 출석에서 필요로 할 때 단위 테스트를 작성하지 않고 구현 먼저 하게 되는 실수를 하기도 했습니다.
- 의식적으로 계속 TDD 싸이클을 따르려고 더 노력해야겠지만, 이런 경우 발견했을 때 테스트 코드를 추가하는 것도 괜찮을까요?
- [test: 캠퍼스 운영 시간 단위 테스트 추가](https://github.com/ljhee92/java-attendance/commit/4be8300e81bf002d82aa4a25b62826a196b47419)
- [test: 교육 시간 단위 테스트 추가](https://github.com/ljhee92/java-attendance/commit/b073b76512e08ec00e1c523b926876019521ee0e)

### 데이터 초기화 책임 분리
- 1단계에서는 파일을 읽은 뒤 ``controller``에서 도메인으로 변환해주었는데, 이번 단계에서는 ``controller``는 '사용자'의 뷰와 도메인을 연결하는 역할이라고만 생각하게 되었습니다.
- 따라서 파일에서 받은 데이터를 초기화하고, 시스템의 날짜를 초기화하는 데이터 초기화의 책임을 ``Initializer``에 위임하였고, 프로그램이 실행될 때 ``Initializer``가 실행되는 것이므로 ``Application``과 같은 단계에 두었습니다.

### 뷰와 도메인 분리
- 1단계에서 피드백으로 뷰와 도메인을 분리해보는 게 어떠냐고 하셔서 1단계에서는 반영해두었습니다.
- 2단계에서는 '구현 요구사항을 만족하는 최소한의 설계를 한다.' 요구사항에 따라 우선 뷰에서 도메인을 받아 출력하도록 구현했으니 참고 부탁드립니다!

### 시스템 날짜
- 2024년 12월 한 달 동안 사용할 출석 시스템을 구현하라는 요구사항에 따라 ``Initializer``에서 시스템 날짜를 2024년 12월로 초기화해주고, 테스트 케이스는 모두 2024년 12월 기준으로 작성했는데요.
- 확장성을 고려해 시스템 날짜가 2025년 2월로 변경되었을 때 프로덕션 코드는 ``Initializer``에서 주입해주는 부분만 변경하면 되겠지만,  1단계에서 샐리가 말씀해주셨던 것처럼 2024년으로 작성된 테스트 케이스는 모두 실패할텐데요.
- 인터페이스를 도입해 시스템 날짜를 주입해주고, 테스트에서도 테스트만의 날짜를 주입해주는 방법을 생각해보았는데 우선 12월로 테스트 케이스를 작성하고 인터페이스 도입하는 방법으로 리팩토링하는 것도 괜찮은 방법일까요?
- 처음부터 인터페이스를 도입하려고 생각하니 복잡하게 느껴져서 우선은 12월 기준으로만 먼저 작성했습니다.

### 테스트 메서드명과 변수명
- 이것도 회바회 팀바팀이겠지만 메서드명 표기를 주로 어떤 방식으로 하시는지 궁금합니다.
  - 메서드명만 한글로 
  - 메서드명은 영어로, ``DisplayName``을 활용하여 한글로 
  - 메서드명도, ``DisplayName``도 한글로
- 저는 다 사용해보긴 했는데, 메서드명만 한글로 사용하는 방식이 시간상 효율적이라고 생각되어 사용하고 있습니다.
- 또, 변수명의 경우 red 단계에서 의미없는 변수명이어도 괜찮은지도 궁금합니다. 어차피 리팩토링을 거치면서 테스트 케이스를 하나로 합치는 경우 의미없는 변수명이 사라지기는 할텐데, 괜히 작성하면서 '이렇게 작성해도 될까..😅' 하는 마음이 듭니다.
   ```java
   @Test
   void 월요일_1시부터_1시5분까지는_출석이다() {
       LocalDate monday = LocalDate.of(2024, 12, 9);
       LocalTime mondayEducationTime = LocalTime.of(13, 0);
       LocalTime mondayEducationTime2 = LocalTime.of(13, 5); // 이러한 의미 없는 변수명...
   
       assertAll(() ->
               assertThat(new Attendance(monday, mondayEducationTime).determineStatus()).isEqualTo("출석"),
               assertThat(new Attendance(monday, mondayEducationTime2).determineStatus()).isEqualTo("출석")
       );
   }
   ```

ljhee92 committed:
chore: 기존 코드 전부 삭제

ljhee92 committed:
docs: 요구사항 명세 정리

ljhee92 committed:
chore: .gitkeep 파일 삭제

ljhee92 committed:
chore: attendances.csv 파일 추가

ljhee92 committed:
test(red): 출석 유효성 검증 테스트(주말)

ljhee92 committed:
feat(green): 출석 유효성 검증 구현(주말)

ljhee92 committed:
refactor: 공통 사용 예외 메시지 Enum 분리

- 추후 운영 시간 예외 메시지도 포맷팅할 가능성을 고려하여 FormattedErrorMessage로 분리했습니다.

ljhee92 committed:
test(red): 출석 유효성 검증 테스트(공휴일)

ljhee92 committed:
feat(green): 출석 유효성 검증 구현(공휴일)

ljhee92 committed:
refactor: 여러 공휴일 테스트할 수 있도록 리팩토링

ljhee92 committed:
refactor: 주말 검증 로직을 DateUtil 클래스로 분리, 주말 테스트 케이스 추가

- 가독성을 위해 주말 검증 로직을 DateUtil 클래스로 분리했습니다.

- 여러 주말을 테스트할 수 있도록 테스트 코드를 리팩토링했습니다.

ljhee92 committed:
docs: 구현한 기능 체크(주말, 공휴일 출석 불가)

ljhee92 committed:
test(red): 출석 유효성 검증 테스트(캠퍼스 운영시간)

ljhee92 committed:
feat(green): 출석 유효성 검증 구현(캠퍼스 운영시간)

ljhee92 committed:
refactor: 23시 00분 59초까지 운영 시간 반영

- 23시 00분 59초까지 운영 시간이므로 23시 1분에 종료된다는 의미로 변수명을 변경했습니다.

- CampusOperatingTime 상수를 활용해 테스트를 리팩토링했습니다.

ljhee92 committed:
docs: 구현한 기능 체크(캠퍼스 운영 시간 출석 불가)

ljhee92 committed:
test(red): 월요일 출석시간 내 출석 시 출석 상태 반환 테스트

ljhee92 committed:
feat(green): 월요일 출석시간 내 출석 시 출석 상태 반환 구현

ljhee92 committed:
refactor: EducationTime 간소화, AttendanceStatus Enum 추가

- 출석, 지각, 결석 시간 만으로도 교육 시간 계산이 되므로 EducationTime을 간소화했습니다.

- 출석 상태를 Enum으로 관리하도록 리팩토링했습니다.

- 중복되는 테스트 코드를 제거하도록 리팩토링했습니다.

ljhee92 committed:
test(red): 월요일 출석시간별 출석 상태 반환 테스트

ljhee92 committed:
feat(green): 월요일 출석시간별 출석 상태 반환 구현

ljhee92 committed:
test(red): 월요일이 아닌 요일 출석시간별 출석 상태 반환 테스트

ljhee92 committed:
feat(green): 월요일이 아닌 요일 출석시간별 출석 상태 반환 구현

ljhee92 committed:
refactor: EducationTime 시간 비교 공통 메서드 분리

ljhee92 committed:
docs: 구현한 기능 체크(교육 시간에 따라 출석 상태 결정)

ljhee92 committed:
test: 캠퍼스 운영 시간 단위 테스트 추가

ljhee92 committed:
test: 교육 시간 단위 테스트 추가

ljhee92 committed:
style: import 정리

ljhee92 committed:
fix: 5분 초과 시 지각, 30분 초과 시 결석 로직 수정(캠퍼스 운영 시간과 혼동)

- 캠퍼스 운영 시간의 경우 23시 1분이 되기 전까지 출석이 가능한데, 교육 시간은 5분이 초과하면 지각, 30분이 초과하면 결석이므로 혼동한 로직을 바로 잡았습니다.

ljhee92 committed:
test(red): 출석 상태 결정 책임 AttendanceStatus에 위임 테스트

- 날짜와 시간에 따라 출석인 상태를 결정하는 책임을 Attendance에서 AttendanceStatus에 위임하기 위해 테스트를 추가했습니다.

ljhee92 committed:
feat(green): 출석 상태 결정 책임 AttendanceStatus에 위임 구현

ljhee92 committed:
refactor: 중복되는 메서드명 수정

- EducationTime에서는 출석 종료 시간 이전인지, 지각 시간 사이인지, 결석 시작 시간 이후인지 판단하는 것이므로 메서드명과 테스트 메서드명을 수정했습니다.

ljhee92 committed:
test(red): 징계 판단 테스트

ljhee92 committed:
feat(green): 징계 판단 구현

ljhee92 committed:
refactor: 징계 테스트 중복 제거 및 테스트 케이스 추가

ljhee92 committed:
docs: 구현한 기능 체크(징계 판단)

ljhee92 committed:
style: import 및 라인 정리

ljhee92 committed:
test(red): 닉네임 검증 테스트

ljhee92 committed:
feat(green): 닉네임 검증 구현

- 크루의 닉네임은 값 객체(VO)라고 생각하기 때문에 record로 구현했습니다.

ljhee92 committed:
test(red): 크루 리스트 추가, 중복 검증 테스트

ljhee92 committed:
feat(green): 크루 리스트 추가, 중복 검증 구현

ljhee92 committed:
test(red): 크루 리스트에 포함된 크루인지 검증 테스트

ljhee92 committed:
feat(green): 크루 리스트에 포함된 크루인지 검증 구현

ljhee92 committed:
feat: 파일 내용 읽기 구현

ljhee92 committed:
test(red): 출석 일자 동일 여부 테스트

ljhee92 committed:
feat(green): 출석 일자 동일 여부 구현

ljhee92 committed:
test(red): 출석 리스트 추가, 중복 검증 테스트

ljhee92 committed:
feat(green): 출석 리스트 추가, 중복 검증 구현

ljhee92 committed:
docs: 구현한 기능 체크(출석 중복 확인)

ljhee92 committed:
test(red): 출석부에 크루와 출석리스트 추가 테스트

ljhee92 committed:
feat(green): 출석부에 크루와 출석리스트 추가 구현

ljhee92 committed:
fix: 크루 추가 로직 수정

- 크루가 추가될 때 동일한 크루가 크루 리스트에 있다면 예외를 던지지 않고 덮어쓰도록 수정했습니다.

ljhee92 committed:
feat: 필요한 정보 초기화 구현

- step1에서 Controller가 했던 초기화 기능을 Initializer에게 부여했습니다.

ljhee92 committed:
docs: 구현한 기능 체크(파일 입력 및 정보 초기화)

ljhee92 committed:
test(red): 출석리스트 출석 수정 테스트

ljhee92 committed:
feat(green): 출석리스트 출석 수정 구현

ljhee92 committed:
test(red): 출석부에서 크루의 출석리스트 수정 테스트

ljhee92 committed:
feat(green): 출석부에서 크루의 출석리스트 수정 구현

ljhee92 committed:
refactor: 테스트에서만 사용되는 size() 메서드 제거

- assertThatCode로 정상코드인지 검증하고 테스트에서만 사용되는 size() 메서드를 제거했습니다.

ljhee92 committed:
test(red): 크루 검증 테스트

ljhee92 committed:
feat(green): 크루 검증 구현

ljhee92 committed:
docs: 구현한 기능 체크(크루 검증)

- 크루 검증 외에 기존 구현된 기능(주말 및 공휴일 검증, 캠퍼스 운영 시간 검증)도 체크했습니다.

ljhee92 committed:
feat: 시스템 초기화 구현

- 시스템 실행 필요한 파일 데이터 입력과 시간 설정을 Initializer 에서 수행하도록 구현했습니다.

ljhee92 committed:
feat: 기능 입력, 출석 확인 기능 구현

ljhee92 committed:
feat: 출석 수정 기능 구현

ljhee92 committed:
test(red): 입력 일자와 출석의 일자, 연월 동일 여부 테스트

- 입력 일자와 출석의 일자 동일 여부 테스트는 출석리스트 출석 수정 구현(커밋 c1a77c7f) 시 테스트를 먼저 구현하지 못하여 추가했습니다.

ljhee92 committed:
feat(green): 입력 일자와 출석의 연월 동일 여부 구현

ljhee92 committed:
test(red): 출석리스트에서 이번 달 전 날까지의 기록 반환 테스트

ljhee92 committed:
feat(green): 출석리스트에서 이번 달 전 날까지의 기록 반환 구현

ljhee92 committed:
test(red): 출석부에서 해당날짜 이전 크루 출석 기록 조회 테스트

ljhee92 committed:
feat(green): 출석부에서 해당날짜 이전 크루 출석 기록 조회 구현

ljhee92 committed:
feat: 크루별 출석 기록 확인 기능 구현

- 전날까지의 기록 확인 부분만 구현
- 출석, 지각, 결석 횟수와 대상자 여부는 구현 필요

ljhee92 committed:
test(red): 출석 상태 통계 반환 테스트

ljhee92 committed:
feat(green): 출석 상태 통계 반환 구현

ljhee92 committed:
feat: 크루별 출석 기록 확인 기능 구현

ljhee92 committed:
test(red): 출석부에서 크루와 출석 통계 반환 테스트

ljhee92 committed:
feat(green): 출석부에서 크루와 출석 통계 반환 구현

ljhee92 committed:
feat: 제적 위험자 확인 기능 구현

ljhee92 committed:
feat: 제적 위험자 정렬 구현

ljhee92 committed:
refactor: 예외시 재입력 처리, 매직넘버 처리, 중복 코드 메서드 분리

ljhee92 committed:
fix: 출력값 오류 수정

ljhee92 committed:
refactor: 사용하지 않는 객체 제거

ljhee92 committed:
fix: 시스템 날짜 수정, 주말 공휴일 검증 위치 추가

ljhee92 committed:
docs: 구현 기능 체크

ljhee92 committed:
refactor: 옵션 선택 Runnable 활용, controller 내 주말/공휴일 검증 매개변수 추가

- 옵션이 늘어날 경우를 고려해 if문 -> Runnable 활용하도록 수정했습니다.
- 출석 수정 시 시스템 날짜를 검증하고 있었으므로 수정하려는 날짜를 검증할 수 있도록 수정했습니다.

ljhee92 committed:
refactor: InputView 에러 메시지 분리, 등교 시간 형식 검증 View에서 처리

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,122 +1,119 @@
-                DateTimeFormatter.ofPattern(OutputView.DATE_FORMATTER, Locale.KOREAN))));
+    private void execute(String inputFunction) {
+        Runnable runnable = option.getOrDefault(inputFunction, null);
+        if (runnable == null) {
```
null을 사용하는걸 저는 꺼려하는 편이에요.
null로 인해 발생할 수 있는 상황들이 추적하기 어렵거든요.
다른 방안으로 풀어내볼까요?

ljhee92 replied:
저도 웬만하면 NPE를 피하기 위해서 null 사용을 지양하려고 하는데 급하다보니 null을 사용하고 리팩토링을 하지 못했네요.
Option을 Enum으로 상수화하는 방안으로 풀어내보았습니다.

[refactor: null 제거, 선택 옵션 Enum으로 분리, 옵션별 기능 Map 초기화 방식 변경](https://github.com/ljhee92/java-attendance/commit/78efc3480b78ffefd49f8f9ea587928c7d60f964)

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,122 +1,119 @@
+        option.put(RECORD, this::recordAttendance);
+        option.put(EDIT, this::editAttendance);
+        option.put(CHECK_RECORD, this::checkRecords);
+        option.put(CHECK_PENALTY, this::checkPenalty);
```
이 option들을 가진 map을 애초에 생성해주면 되지 않나요?
initialize의 이유가 궁금해요.

ljhee92 replied:
``Map.of()``를 사용해서 애초에 생성해주는 방식을 생각해내지 못해서 ``initialize`` 메서드를 만들었습니다.
애초에 불변의 map을 생성해주는 것이 더 안정적인 방식이겠네요! 적용해두었습니다 ☺️

[refactor: null 제거, 선택 옵션 Enum으로 분리, 옵션별 기능 Map 초기화 방식 변경](https://github.com/ljhee92/java-attendance/commit/78efc3480b78ffefd49f8f9ea587928c7d60f964)

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,122 +1,119 @@
-        Attendance modifiedAttendance = attendancesBook.modify(crew, previousAttendance, modifyingCheckInTime);
-        OutputView.printModifyingResult(previousAttendance, modifiedAttendance);
+    private Crew getCrew(String inputNickname) {
+        Crew crew = new Crew(new Nickname(inputNickname));
```
```suggestion
        Nickname nickname = new Nickname(inputNickname);
        nickname.validateBlankOrNull(nickname);

        Crew crew = new Crew(nickname);
```
저는 이런 식의 코드를 선호하는 편이에요.
객체를 생성하면서 내부에서 validation이 들어갈 것이라고 기대하지 않아서인 것 같아요.

ljhee92 replied:
```java
Nickname nickname = new Nickname(inputNickname);
Crew crew = new Crew(nickname);
```
위와 같이 닉네임 생성을 명시하고 크루에 주입해주는 것에는 동의합니다.
다만 ``validateBlankOrNull`` 을 생성 시에 검증하지 않고 public으로 열게 되면 생성하는 곳에서 매 번 validation을 호출해주어야 하는걸까요?

dusdn1702 replied:
저는 그걸 선호하는데요~! 모든 생성에서 유효성 검증이 필요하다면 생성자 내부에 넣을 수도 있을 것 같아요. 😄

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,122 +1,119 @@
-        Attendance previousAttendance = attendancesBook.getExistAttendanceOfCrew(crew, modifyingCheckInDate);
-        Attendance modifiedAttendance = attendancesBook.modify(crew, previousAttendance, modifyingCheckInTime);
-        OutputView.printModifyingResult(previousAttendance, modifiedAttendance);
+    private Crew getCrew(String inputNickname) {
```
getCrew 보다는 createCrew 같은 함수네요!

ljhee92 replied:
동의합니다! ``inputNickname``을 받아 ``Crew``를 반환하는 메서드이니 ``createCrew``가 더 적합하겠어요!

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,87 @@
+            return;
+        }
+
+        attendanceBook.put(crew, new Attendances(new ArrayList<>(List.of(attendance))));
```
new ArrayList<>가 들어가야했던 이유가 뭘까요?

ljhee92 replied:
``List.of(attendance)``만 주입하면 출석 리스트가 불변 리스트로 생성되므로 크루별 출석 기록을 추가하거나 수정할 수 없어서 ``new ArrayList<>()``로 주입해주었습니다.
map에 ``putIfAbsent`` 메서드를 활용하면 크루가 존재하지 않을 경우 빈 출석 리스트를 넣어주게 할 수 있어서 아래와 같이 개선했습니다.
```java
public void add(Crew crew, Attendance attendance) {
    attendanceBook.putIfAbsent(crew, new Attendances(new ArrayList<>()));
    attendanceBook.get(crew).add(attendance);
}
```

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,87 @@
+
+    public Attendance findAttendanceByCrew(Crew crew, LocalDate inputDate) {
+        return attendanceBook.get(crew).findByDate(inputDate);
+    }
```
```suggestion
    public Attendance findAttendanceByCrewAndDate(Crew crew, LocalDate inputDate) {
        Attendances attendances = attendanceBook.get(crew);
        return attendances.findByDate(inputDate);
    }
```
저라면 가독성을 위해 이렇게 할 것 같아요~

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,87 @@
+        attendances.update(oldAttendance, newAttendance);
+    }
+
+    public void validateCrew(Crew crew) {
```
유효성도 일관된 위치에서 수행하면 좋을 것 같아요.
어디서는 생성자 내부에서 수행하고, 어디서는 함수로 빼서 수행하면 관리하기 어렵지 않을까요?

ljhee92 replied:
``AttendanceBook``에서 수행하고 있는 ``validateCrew`` 메서드는 출석부에 포함되어 있는지 검증하는 것이기 때문에 ``AttendanceBook`` 내부에서 수행하고 있다고 생각해요.
샐리의 피드백을 듣고 보니 메서드명의 네이밍이 잘못되었다고 생각되네요. ``validateCrew``보다는 ``validateContainsCrew`` 정도가 적합할 것 같아 변경해두었습니다.

[refactor: 출석부에 크루가 없다면 예외를 던지는 유효성 검증 메서드명 변경](https://github.com/ljhee92/java-attendance/commit/8672d4c9cd5b1db8bfa55d6812f5c81afc0ad95c)

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,87 @@
+        return penaltyCrews;
+    }
+
+    private int compareCrew(LocalDate today, Crew crew, Crew otherCrew) {
```
이건 무슨 함수인가요? 결석을 많이한 순서로 정렬하고 같으면 이름순인건가요?

ljhee92 replied:
네 맞습니다!
크루별로 총 결석횟수에 따라 정렬하고 같으면 이름순으로 정렬하는 메서드입니다.

dusdn1702 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -17,87 +14,42 @@ public Attendances(List<Attendance> attendances) {
-        return attendances.stream()
-                .filter(attendance -> attendance.getAttendanceDateTime().toLocalDate() == modifyingDate)
-                .peek(attendance -> attendance.modify(LocalDateTime.of(modifyingDate, modifyingTime)))
+    public void update(Attendance oldAttendance, Attendance newAttendance) {
```
기존 시간을 변경하는 방식에서 출석 객체 자체를 변경해주신 이유가 궁금해요!
update 전에 findAttendanceByCrew로 가져온 oldAttendance인데 stream filtering을 거칠 필요가 있을까요?

ljhee92 replied:
기존 시간을 변경하는 방식에서 출석 객체 자체를 변경한 이유는 기존처럼 시간을 변경하게 되면 ``Attendance`` 객체의 불변이 보장되지 않기 때문입니다.
기존 방식에서는 ``Attendance`` 객체와 ``Attendances``  컬렉션 객체의 불변성이 모두 보장되지 않는데, 출석 객체 자체를 변경하게 되면 여전히 ``Attendances`` 컬렉션은 변경되지만 ``Attendance`` 객체 자체는 불변성을 보장할 수 있게 되기 때문에 변경했습니다.

샐리 말씀대로 stream filtering 거치지 않고 ``oldAttendance``의 index만 찾아와서 변경해주어도 되겠네요! 코드 구현할 때 한 번만 더 단순하게 생각해보았으면 깨달았을 것 같은데 너무 매몰되어 있었던 것 같습니다. 변경해두었습니다!

[refactor: 출석 수정 시 stream filtering 제거](https://github.com/ljhee92/java-attendance/commit/04aa874e66fd2046a3ad936d144690d2c09de62a)

dusdn1702 commented on src/main/java/attendance/domain/CampusOperatingTime.java:
```diff
@@ -0,0 +1,24 @@
+public enum CampusOperatingTime {
+
+    OPEN_AT(LocalTime.of(8, 0)),
+    CLOSE_AT(LocalTime.of(23, 1))
```
23:01 이 닫는 시간인가요~? isBefore을 사용하기 위해 이렇게 작성했다면 요구사항과 맞는 코드를 작성해보면 좋을 것 같아요.
이 프로그램이 추후 확장되어서 초 단위까지 관리하게 된다면 이 부분이 문제가 되지 않을까요?

ljhee92 replied:
제가 캠퍼스 운영 시간과 교육 운영 시간을 나중에 테스트하다 보니 이런 혼동이 있었던 것 같아요😭
지금 다시 생각해보니 요구사항인 '캠퍼스 운영 시간은 매일 08:00~23:00이다.' 를 잘못해석했었네요.
'23시 00분 59초여도 출석이 되어야 하는 게 아닐까?'라고 생각했는데, 23:00까지라면 23:00:00까지만 가능한 게 맞네요!
다시 한 번 TDD의 중요성을 느끼고.. 수정해두었습니다🥹

[fix: 23시 초과 시 출석 불가 로직 수정](https://github.com/ljhee92/java-attendance/commit/6f2ca429e42f0cc2b9fcf7c5231920282ef591ff)

dusdn1702 commented on src/main/java/attendance/domain/Crew.java:
```diff
@@ -4,27 +4,25 @@
 
 public class Crew {
```
Crew 객체의 역할은 무엇인가요? 내부에서 수행하고 있는 함수가 하나도 없다는 것은 책임이 없는 무의미한 객체 아닐까요?

ljhee92 replied:
현재로서는 내부에서 수행하고 있는 함수가 하나도 없어 무의미한 객체라고 생각할 수 있겠지만,
요구사항에서 지속적으로 닉네임이 아니라 크루의 출석, 크루의 출석 기록 확인이라고 명명하고 있었기 때문에 확장 가능성을 염두에 두고 크루 객체를 생성했어요!

dusdn1702 commented on src/main/java/attendance/domain/Nickname.java:
```diff
@@ -0,0 +1,21 @@
+
+import attendance.util.ErrorMessage;
+
+public record Nickname(String nickname) implements Comparable<Nickname> {
```
다른 클래스들도 record 성격을 띄고 있는 클래스들이 꽤 보이는데 Nickname만 record 선언해주신 이유가 있나요?

ljhee92 replied:
주로 record는 값 객체나 DTO에 활용된다는 것만 알고 있어서 ``Nickname`` 객체만 값 객체(VO)라고 생각하여 record로 선언했는데요.
record의 성격을 찾아보니 불변 객체도 해당하는 것을 알게 됐어요!
``Attendance`` 객체도 현재 불변 객체로 설계했기 때문에 ``Attendance`` 객체도 record 로 변경해주었습니다.

[refactor: 불변 객체인 Attendance record로 변경](https://github.com/ljhee92/java-attendance/commit/948fa9ce4db3793a3b534e630de77ffaf649db86)

ljhee92 replied:
다만 추후 리팩토링을 진행하다보니 Attendance를 정적 팩토리 메서드로 생성하게 되어 record의 경우 생성자를 private으로 막을 수 없어 다시 일반 class로 되돌렸습니다.😅

dusdn1702 commented on src/main/java/attendance/domain/StatusStatistics.java:
```diff
@@ -0,0 +1,45 @@
+
+    private final Map<AttendanceStatus, Integer> statusStatistics;
+
+    public StatusStatistics(List<Attendance> attendances, LocalDate today) {
```
Attendances 객체를 사용하지 않은 이유가 있을까요?

ljhee92 replied:
일급 컬렉션 활용이 익숙하지 않아서 무의식적으로 익숙한 ``List<Attendance>``를 반환하여 사용했습니다.
``List<Attendance>``가 사용되던 부분을 모두 ``Attendances`` 객체를 사용하는 것으로 변경해두었습니다.
변경사항이 조금 많은데 변경한 이유들을 커밋 로그에 상세히 기록해두었습니다!

[refactor: List<Attendance> -> Attendances로 변경하여 중복 제거](https://github.com/ljhee92/java-attendance/commit/4c9630b1e8f53d855fcaf0935422492143ba9b70)

dusdn1702 commented on src/main/java/attendance/domain/StatusStatistics.java:
```diff
@@ -0,0 +1,45 @@
+
+    private void calculateStatistics(Map<AttendanceStatus, Integer> statistics, List<Attendance> attendances, LocalDate currentDate) {
+        AttendanceStatus status = attendances.stream()
+                .filter(attendance -> attendance.isSameDate(currentDate))
```
Attendances#findByDate 가 많이 활용될 수 있는데 중복 코드가 많이 보이는 것 같아요.

ljhee92 replied:
이 부분도 ``List<Attendance>``를 사용하여 중복 코드가 많이 발생한 것 같다고 생각이 듭니다.
``Attendances`` 객체를 활용하는 것으로 코드를 바꾸면서 적용해두었습니다.

dusdn1702 commented on src/main/java/attendance/util/FileResourceReader.java:
```diff
@@ -0,0 +1,23 @@
+
+public class FileResourceReader {
+
+    private static final int HEADER = 1;
```
👍

dusdn1702 commented on src/test/java/attendance/domain/AttendanceStatusTest.java:
```diff
@@ -1,20 +1,70 @@
-    @ParameterizedTest
-    void 출석일자와_시간을_입력하면_맞는_상태를_반환한다(String dateTime, AttendanceStatus expectedStatus) {
-        LocalDateTime attendanceDateTime = LocalDateTime.parse(dateTime);
+    @ParameterizedTest(name = "{index} : {2}")
```
{index} : {2} 요게 뭔가요??

ljhee92 replied:
``@ParameterizedTest``의 ``@MethodSource``를 사용할 때 활용할 수 있는 출력 방식으로 알고 있습니다.
테스트명을 출력할 때 {index}는 1부터 출력해주고, {2}는 파라미터 중 세 번째 파라미터(0부터 카운트)인 message를 출력해줍니다.
[ParameterizedTest 공부](https://dublin-java.tistory.com/56)하면서 알게 됐어요!

<img width="997" alt="스크린샷 2025-03-03 오후 3 25 04" src="https://github.com/user-attachments/assets/61c93a66-3b30-448a-8ed5-90648cf81554" />
<img width="275" alt="스크린샷 2025-03-03 오후 3 25 17" src="https://github.com/user-attachments/assets/98d282d1-731b-4dc4-8269-1b5312708597" />

dusdn1702 replied:
아하 저도 몰랐네요 ㅎㅎ

dusdn1702 commented on src/test/java/attendance/domain/AttendanceStatusTest.java:
```diff
@@ -1,20 +1,70 @@
-        LocalDateTime attendanceDateTime = LocalDateTime.parse(dateTime);
+    @ParameterizedTest(name = "{index} : {2}")
+    @MethodSource("getAttendDateAndTime")
+    void 날짜와_시간에따라_출석인_상태를_반환한다(LocalDate inputDate, LocalTime inputTime, String message) {
```
message들은 다 미사용이네요!

ljhee92 replied:
메서드 내에서는 미사용이지만 위에서 답변드린 것처럼 테스트명 출력 시 사용되고 있는데요,
네이밍을 바꿔보는 것이 좋을까요?!🤔

dusdn1702 replied:
테스트명을 이렇게 출력하는게 추후 테스트 확장, 유지보수에 좋을지에 대해 고민해보면 좋을 것 같아요!

dusdn1702 commented on src/test/java/attendance/domain/AttendanceTest.java:
```diff
@@ -1,95 +1,160 @@
-        LocalDateTime attendanceDateTime = LocalDateTime.of(weekend, attendanceTime);
+    @ParameterizedTest(name = "{index} : {1}")
+    @MethodSource("getWeekend")
+    void 주말에_출석을_하면_예외가_발생한다(LocalDate weekend, String message) {
```
💯 👍

dusdn1702 commented on src/main/java/attendance/Initializer.java:
```diff
@@ -0,0 +1,43 @@
+import java.time.format.DateTimeFormatter;
+import java.util.HashMap;
+
+public class Initializer {
```
Initializer의 위치에 대해 고민해보면 좋을 것 같아요. 실행될 때 실행되는 것은 맞지만 내부에서 도메인을 생성하고 있는데 Application 만큼 외부에 존재하는게 좋을까요?

ljhee92 replied:
단순히 파일 시스템을 통해서 프로그램 실행 시점에 객체를 초기화하여 주입해준다고만 생각해서 위치를 정했는데 도메인을 내부에서 생성하고 있다는 점에 집중하니 ``Application``과 같은 위치에 존재하면 안 되겠다고 생각이 드네요.
별도의 ``infrastructure`` 패키지를 생성하여 위치를 변경해주었습니다.

[chore: Initializer 위치 변경](https://github.com/ljhee92/java-attendance/commit/c9c2b913c3ad7c210be293e73e25e33418d52399)

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,87 @@
+        int absentCount = statistics.getAttendanceStatusCount(AttendanceStatus.ABSENT);
+        int lateCount = statistics.getAttendanceStatusCount(AttendanceStatus.LATE);
+
+        return Penalty.calculateTotalAbsent(lateCount, absentCount);
```
왜 calculateTotalAbsent을 static으로 사용하게 되었을까요? 결석 계산의 책임을 어디서 갖는게 좋을까요?

ljhee92 replied:
``Penalty`` 객체가 제적/경고/면담을 결정하는 책임을 가지고 있고, 궁극적으로 그 결정을 위해 필요한 결석 계산이기 때문에 ``Penalty`` 객체가 결석 계산도 담당하도록 했습니다.
샐리의 질문을 듣고 나니 결석 계산 자체에 대한 책임이라고 생각하면 ``Attendances``가 그 책임을 갖는 것이 더 맞지 않을까? 라는 생각도 드는데요, static으로 사용했을 경우 문제점이 있을까요?

dusdn1702 replied:
객체 지향을 공부하고 있는 상황에서 static 함수는 그 의미가 퇴색된다고 생각했어요.
`제적/경고/면담을 결정 하는 것`과 `그 결정을 위한 결석 계산`의 책임에 대해 생각해보면 좋을 것 같아요.

dusdn1702 commented:
commit 단위
- 사실 말씀 주신 것처럼 모두 다르겠지만 저 같은 경우는 프로젝트의 단위가 크기 때문에 커밋을 이렇게 찍어두면 어느 시점에 어떤 히스토리로 수정된 코드인지 파악하기 어렵더라구요. 팀 내 커밋 방식의 규칙을 세우고 적용하는 편입니다. 지금은 실무보다는 이런 싸이클로도 경험해보고, 추후 협업도 해보면서 어떤 커밋 단위가 좋을지 듀이의 기준을 세워 나가면 좋을 것 같아요
TDD
- 기능 단위 TDD를 하다보면 충분히 그럴 수 있다고 생각해요. 필요에 따라 테스트 코드를 놓치지 않고 추가하는 것도 중요하다고 생각해요. 
시스템 날짜
- 좋습니다~! 다만 저라면, Initializer의  initSystemDate를 가져다 사용할 것 같네요 ㅎㅎ
테스트 메서드명
- 메서드명은 영어로 작성하고 displayName에 한글로 작성하는 편이에요. 말씀 주신 것처럼 변경 가능성이 크기 때문에 메서드명에 테스트의 목적을 작성하고 상세한 설명을 displayName에 적으려 합니다.

dusdn1702 reviewed (changes requested):
안녕하세요 듀이 ~

커밋 기록을 보니 기능을 하나하나 구현해가면서 TDD를 잘 수행해주신 것 같아요.
1단계 반영해주신 것도 확인했습니다 ㅎㅎ 👍 같은 프로그램을 이렇게 다시 구현해보니 어떠셨나요? 좋은 경험이 되셨길 바라요!

질문에 대한 답변과 몇가지 궁금한 점 코멘트 남겨 두었어요. 확인 부탁드릴게요 😄

ljhee92 committed:
refactor: null 제거, 선택 옵션 Enum으로 분리, 옵션별 기능 Map 초기화 방식 변경

- null을 제거하는 방안으로 선택 옵션을 Enum으로 분리했습니다.
- 옵션별 기능을 포함하는 Map을 바로 생성하는 방식으로 변경했습니다.

ljhee92 committed:
chore: Initializer 위치 변경

- 파일 입출력(외부)을 통해 내부에서 도메인을 생성하고 있으므로 infrastructure 로 변경했습니다.

ljhee92 committed:
refactor: Nickname 중간 객체 분리

ljhee92 committed:
refactor: inputNickname을 받아 Crew를 생성하는 메서드이므로 getCrew -> createCrew로 메서드명 변경

ljhee92 committed:
refactor: 가독성을 위해 Attendances 중간 객체 분리

ljhee92 committed:
refactor: 출석부에 크루가 없다면 예외를 던지는 유효성 검증 메서드명 변경

ljhee92 committed:
refactor: 출석 수정 시 stream filtering 제거

ljhee92 committed:
fix: 23시 초과 시 출석 불가 로직 수정

- 캠퍼스 운영 시간이 08:00~23:00 라면 23:00:01이 되는 순간부터 출석이 불가한 것이므로 로직과 상수를 수정했습니다.

ljhee92 committed:
refactor: 불변 객체인 Attendance record로 변경

ljhee92 committed:
refactor: List<Attendance> -> Attendances로 변경하여 중복 제거

- Attendances의 findByDate 메서드를 활용할 수 있으므로 List<Attendance>가 사용되었던 곳들을 Attendances를 사용하도록 변경했습니다.

- 기존 findByDate 메서드에서 날짜로 찾는 출석 객체가 없으면 예외를 발생 시켰는데, 예외를 발생시키지 않고 날짜만으로 출석 객체를 생성하도록 변경했습니다.

- 날짜와 시간으로 Attendance를 생성하는 경우 날짜, 시간 모두 유효성 검증을 하고 있었기 때문에 날짜만으로 Attendance를 생성할 때에는 날짜만 검증하고 시간은 결석 고정값(00:00)으로 생성하도록 정적 팩토리 메서드를 사용했습니다.

- 정적 팩토리 메서드 사용 시 record를 유지한다면 생성자를 private으로 막을 수 없어 class로 다시 변환했습니다.

- 날짜만으로 Attendance 객체 생성 시 시간을 결석 고정값(00:00)으로 두었더니 AttendanceStatus 결정 로직에서 운영시간 전이어도 출석으로 정해버리는 버그가 있어 해결했습니다.

ljhee92 committed:
refactor: putIfAbsent method를 활용하여 크루가 출석부에 존재하지 않을 경우 빈 출석 리스트 생성한 뒤 추가

ljhee92 commented:
샐리 안녕하세요!
휴일에도 좋은 리뷰와 피드백 감사드립니다😄

덕분에 일급 컬렉션을 좀 더 적극적으로 활용해보도록 노력하는 시간이 되었어요.
같은 프로그램을 이렇게 다시 구현하니 굉장히 색다른 경험이었어요.
처음엔 도메인에 대해 다 알고 있다고 생각해서 금방 구현할 거라고 생각했는데,
TDD를 적용하며 구현하다 보니 오히려 알고 있는게 독이 아닌가 라는 생각도 들었네요😅

TDD를 엄격하게 지키려고 하다보니 과정 중에는 정말 힘들었던 기억밖에 없는 것 같은데
막상 리팩토링을 하려니까 테스트가 굉장히 든든함을 주어서 신기했어요.
코멘트 남겨주셨던 부분들에 저도 동의하는 부분은 반영하여 커밋 기록과 함께 남겨두었고,
궁금한 점이 있는 부분은 질문 남겨두었으니 확인해주시면 좋겠습니다!

커밋 내역도 많고 바뀐 부분이 많아 힘드실텐데도 꼼꼼하게 리뷰해주시고 중간 중간 칭찬 코멘트도 남겨주셔서 감사합니다🥹👍

ljhee92 committed:
fix: 출석 기록이 없는 출석을 수정하려고 할 때 버그 수정

- 출석 기록이 없는 출석을 수정하려고 할 때 시간이 00:00인 객체를 생성하도록 했으므로 index를 찾을 수 없어 발생하는 버그 수정

ljhee92 commented:
앗 샐리 죄송해요 리뷰요청 드리고 마지막에 버그 하나를 발견해서 후다닥 수정해서 푸쉬해두었어요!
리뷰요청하고 푸쉬하는거 아니라고 들었는데 오래 걸리지 않을 것 같아 급하게 수정했습니다 양해 부탁드려요😭😭

dusdn1702 reviewed (approved):
안녕하세요 듀이~

피드백 반영해주신 부분 확인했습니다.
TDD도 열심히 수행해주시고 처음보다 객체지향적으로 리팩터링 잘 진행해주셔서 많이 깔끔해졌네요 😄 

이번 미션은 여기서 마무리 짓도록 할게요. 다음 미션도 화이팅입니다! 🔥
